std/experimental/checkedint.d

```diff
@@ -0,0 +1,1626 @@
+            }
+            else static if (valueConvertible!(T, R) && valueConvertible!(V, R))
+            {
+                return payload == rhs.payload;
+            }
+            else static if (hasMember!(Hook, "hookOpEquals"))
+            {
+                return hook.hookOpEquals(payload, rhs.payload);
+            }
+            else static if (hasMember!(Hook1, "hookOpEquals"))
+            {
+                return rhs.hook.hookOpEquals(rhs.payload, payload);
+            }
+            else
+            {
+                return payload == rhs.payload;
+            }
+        }
+        else static if (hasMember!(Hook, "hookOpEquals"))
+            return hook.hookOpEquals(payload, rhs);
+        else static if (isIntegral!U || isFloatingPoint!U || is(U == bool))
+            return payload == rhs;
+    }
+
+    // opCmp
+    /**
+    */
+    auto opCmp(U)(const U rhs) //const pure @safe nothrow @nogc
+    if (isIntegral!U || isFloatingPoint!U || is(U == bool))
+    {
+        static if (hasMember!(Hook, "hookOpCmp"))
+        {
+            return hook.hookOpCmp(payload, rhs);
+        }
+        else static if (valueConvertible!(T, U) || valueConvertible!(U, T))
+        {
+            return payload < rhs ? -1 : payload > rhs;
+        }
+        else static if (isFloatingPoint!U)
+        {
+            U lhs = payload;
+            return lhs < rhs ? U(-1.0)
+                : lhs > rhs ? U(1.0)
+                : lhs == rhs ? U(0.0) : U.init;
+        }
+        else
+        {
+            return payload < rhs ? -1 : payload > rhs;
+        }
+    }
+
+    /// ditto
+    auto opCmp(U, Hook1)(Checked!(U, Hook1) rhs)
+    {
+        alias R = typeof(payload + rhs.payload);
+        static if (valueConvertible!(T, R) && valueConvertible!(T, R))
+        {
+            return payload < rhs.payload ? -1 : payload > rhs.payload;
+        }
+        else static if (is(Hook == Hook1))
+        {
+            // Use the lhs hook
+            return this.opCmp(rhs.payload);
+        }
+        else static if (hasMember!(Hook, "hookOpCmp"))
+        {
+            return hook.hookOpCmp(payload, rhs);
+        }
+        else static if (hasMember!(Hook1, "hookOpCmp"))
+        {
+            return rhs.hook.hookOpCmp(rhs.payload, this);
+        }
+        else
+        {
+            return payload < rhs.payload ? -1 : payload > rhs.payload;
+        }
+    }
+
+    // opUnary
+    /**
+    */
+    auto opUnary(string op)()
+    if (op == "+" || op == "-" || op == "~")
+    {
+        static if (op == "+")
+            return Checked(this); // "+" is not hookable
+        else static if (hasMember!(Hook, "hookOpUnary"))
+        {
+            auto r = hook.hookOpUnary!op(payload);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else static if (!isUnsigned!T && op == "-" &&
+                hasMember!(Hook, "onOverflow"))
+        {
+            import core.checkedint;
+            alias R = typeof(-payload);
```

Comment:
Actually it's necessary if `payload` is another `Checked` instance that gets creative with unary minus. Now I realize that case would err in `negs` so I'll fix the static if as well.